executor/aggfuncs/func_group_concat.go

```diff
@@ -32,7 +33,7 @@ type baseGroupConcat4String struct {
 	// According to MySQL, a 'group_concat' function generates exactly one 'truncated' warning during its life time, no matter
 	// how many group actually truncated. 'truncated' acts as a sentinel to indicate whether this warning has already been
 	// generated.
-	truncated bool
+	truncated *int32
```

Comment:
can we remove this field, use `sctx.GetSessionVars().StmtCtx.GroupConcatTruncated` instead.

Reply:
No, I found stmtCtx level is bug... fix one stmt with two `group_conat()` functions need got 2 warnings, so `StmtCtx.GroupConcatTruncated` has be removed.

and @spongedu's https://github.com/pingcap/tidb/commit/e3ba021f5a25c3b4ba63fe91f904da270f7fc22a seems not all right..

for mysql, 


```
mysql> select version();
+-----------------------------+
| version()                   |
+-----------------------------+
| 5.7.23-0ubuntu0.18.04.1-log |
+-----------------------------+
1 row in set (0.00 sec)

mysql> select * from tt;
+------+-------+-------+
| id   | a     | b     |
+------+-------+-------+
|    1 | hello | hello |
|    1 | world | world |
|    2 | hello | hello |
|    2 | world | world |
|    3 | hello | hello |
|    3 | world | world |
|    4 | hello | hello |
|    4 | world | world |
|    3 | hello | hello |
|    3 | world | world |
|    4 | hello | hello |
|    4 | world | world |
+------+-------+-------+
12 rows in set (0.01 sec)

mysql> set @@group_concat_max_len=7;                                                                                                                                                               Query OK, 0 rows affected (0.00 sec)

mysql> show warnings;
+---------+------+---------------------------------+
| Level   | Code | Message                         |
+---------+------+---------------------------------+
| Warning | 1260 | Row 2 was cut by GROUP_CONCAT() |
| Warning | 1260 | Row 2 was cut by GROUP_CONCAT() |
| Warning | 1260 | Row 4 was cut by GROUP_CONCAT() |
| Warning | 1260 | Row 4 was cut by GROUP_CONCAT() |
| Warning | 1260 | Row 6 was cut by GROUP_CONCAT() |
| Warning | 1260 | Row 6 was cut by GROUP_CONCAT() |
| Warning | 1260 | Row 8 was cut by GROUP_CONCAT() |
| Warning | 1260 | Row 8 was cut by GROUP_CONCAT() |
+---------+------+---------------------------------+
8 rows in set (0.00 sec)
```

it will issue one warning message for each group and each function call,  so 4 x 2  => 8

but I thinks it's not useful. :rofl:  so keep @spongedu's way for each function call only issue one warning message.

Reply:
what is the query you issued?

Reply:
sorry, miss query...

```
CREATE TABLE `tt` (
  `id` int(11) DEFAULT NULL,
  `a` varchar(100) DEFAULT NULL,
  `b` varchar(100) DEFAULT NULL
) ENGINE=InnoDB;

insert into tt values (1, 'hello', 'hello'), (2, 'world', 'world'), (3, 'hello', 'hello'), (4, 'world', 'world'), (3, 'hello', 'hello'), (3, 'world', 'world'), (4, 'hello', 'hello'), (4, 'world', 'world'), (3, 'hello', 'hello'), (3, 'world', 'world'), (4, 'hello', 'hello'), (4, 'world', 'world');

set @@group_concat_max_len=7; 

select group_concat(a), group_concat(b) from tt group by id;
```

Reply:
It seems that the main code change of this PR is the error/warning message? We still return only one error message per function, ignore the number of rows?

Reply:
As for the "parallel execution problem" described in the PR description, do you mean the data race on the `truncated` field?

Reply:
The main purpose of this pr is make group_concate truncate alway return error in `insert  select`.

before this pr `insert select` with truncated will got 

```
mysql> insert into t1 select group_concat(e) from t2 group by id;
ERROR 1366 (HY000): incorrect utf8 value e58aa0e5(加�) for column p
```
if utf8 value after truncate doesn't incorrect will sliently insert wrong value.

warning count question is meet during test..not main purpose :laughing:

Reply:
No, there are no race condition, 

but hash agg parallel execution will got `x2` warnings..one in partial worker and another in final worker, because each worker alloc seperate  baseGroupConcat4String and `truncated` memory, so duplicate mechanism doesn't work